File: YFDebugPannel/YFDebugPannel/HCEnvPanelViewModel.m
#import "HCEnvPanelViewModel.h"
#import "HCEnvSection.h"
#import "HCCellItem.h"
#import "HCEnvKit.h"
#import "HCEnvBuilder.h"
#import "DependencyEngine.h"
#import "HCPresentationRequest.h"

@interface HCEnvPanelViewModel ()

@property (nonatomic, copy) NSArray<HCEnvSection *> *sections;
@property (nonatomic, strong) DependencyEngine *dependencyEngine;
@property (nonatomic, strong) NSDictionary<NSString *, NSIndexPath *> *indexMap;

@end

@implementation HCEnvPanelViewModel

- (instancetype)init {
    if (self = [super init]) {
        HCEnvSection *envSection = [HCEnvBuilder buildEnvSection];
        _sections = @[envSection];
        [self loadPersistedValues];
        [self rebuildIndexMap];
        [self rebuildDependencyEngine];
        [self refreshAllItems];
    }
    return self;
}

- (void)loadPersistedValues {
    for (HCEnvSection *section in self.sections) {
        for (HCCellItem *item in section.items) {
            if (item.storeKey.length == 0) {
                continue;
            }
            id stored = [[NSUserDefaults standardUserDefaults] objectForKey:item.storeKey];
            if (stored) {
                item.value = stored;
            } else if (item.defaultValue) {
                item.value = item.defaultValue;
            }
            if (item.type == HCCellItemTypeString || item.type == HCCellItemTypeStepper) {
                item.detail = [item.value description];
            }
        }
    }
}

- (void)rebuildDependencyEngine {
    NSMutableArray<HCCellItem *> *items = [NSMutableArray array];
    for (HCEnvSection *section in self.sections) {
        [items addObjectsFromArray:section.items];
    }
    self.dependencyEngine = [[DependencyEngine alloc] initWithItems:items];
}

- (void)refreshAllItems {
    NSDictionary<NSString *, HCCellItem *> *itemsById = self.dependencyEngine.itemsById;
    for (HCEnvSection *section in self.sections) {
        for (HCCellItem *item in section.items) {
            if (item.recomputeBlock) {
                item.recomputeBlock(item, itemsById);
            }
        }
    }
}

- (void)rebuildIndexMap {
    NSMutableDictionary<NSString *, NSIndexPath *> *map = [NSMutableDictionary dictionary];
    [self.sections enumerateObjectsUsingBlock:^(HCEnvSection * _Nonnull section, NSUInteger sectionIndex, BOOL * _Nonnull stop) {
        [section.items enumerateObjectsUsingBlock:^(HCCellItem * _Nonnull item, NSUInteger rowIndex, BOOL * _Nonnull stopRow) {
            if (item.identifier.length > 0) {
                map[item.identifier] = [NSIndexPath indexPathForRow:rowIndex inSection:sectionIndex];
            }
        }];
    }];
    self.indexMap = [map copy];
}

- (HCCellItem *)itemAtIndexPath:(NSIndexPath *)indexPath {
    HCEnvSection *section = self.sections[indexPath.section];
    return section.items[indexPath.row];
}

- (NSArray<NSIndexPath *> *)updateItem:(HCCellItem *)item value:(id)value {
    if (!item) {
        return @[];
    }
    item.value = value;
    if (item.type == HCCellItemTypeString || item.type == HCCellItemTypeStepper) {
        item.detail = [value description];
    }
    if (item.valueTransformer) {
        item.valueTransformer(item);
    }

    NSMutableSet<NSString *> *changed = [NSMutableSet setWithObject:item.identifier];
    if (item.identifier.length > 0) {
        NSSet<NSString *> *propagated = [self.dependencyEngine propagateFromItemId:item.identifier];
        [changed unionSet:propagated];
    }

    [self persistIfNeededForItem:item];
    [self persistEnvConfig];

    NSMutableArray<NSIndexPath *> *paths = [NSMutableArray array];
    for (NSString *itemId in changed) {
        NSIndexPath *indexPath = self.indexMap[itemId];
        if (indexPath) {
            [paths addObject:indexPath];
        }
    }
    return [paths copy];
}

- (void)persistIfNeededForItem:(HCCellItem *)item {
    if (item.storeKey.length == 0) {
        return;
    }
    if (!item.value) {
        [[NSUserDefaults standardUserDefaults] removeObjectForKey:item.storeKey];
    } else {
        [[NSUserDefaults standardUserDefaults] setObject:item.value forKey:item.storeKey];
    }
    [[NSUserDefaults standardUserDefaults] synchronize];
}

- (void)persistEnvConfig {
    NSDictionary<NSString *, HCCellItem *> *itemsById = self.dependencyEngine.itemsById;
    HCCellItem *envItem = itemsById[HCEnvItemIdEnvType];
    HCCellItem *clusterItem = itemsById[HCEnvItemIdCluster];
    HCCellItem *isolationItem = itemsById[HCEnvItemIdIsolation];
    HCCellItem *versionItem = itemsById[HCEnvItemIdVersion];
    if (!envItem || !clusterItem || !isolationItem || !versionItem) {
        return;
    }

    HCEnvConfig *config = [[HCEnvConfig alloc] init];
    config.envType = (HCEnvType)[envItem.value integerValue];
    config.clusterIndex = [clusterItem.value integerValue];
    config.isolation = isolationItem.value ?: @"";
    config.version = versionItem.value ?: @"v1";
    [HCEnvKit saveConfig:config];
}

- (HCPresentationRequest *)presentationForDisabledItem:(HCCellItem *)item {
    if (item.disabledHint.length == 0) {
        return [HCPresentationRequest toastWithMessage:@"当前不可用"];
    }
    return [HCPresentationRequest toastWithMessage:item.disabledHint];
}

@end

File: YFDebugPannel/YFDebugPannel/HCEnvPanelViewModel.swift
import Foundation
import UIKit

final class HCEnvPanelViewModel {
    private(set) var sections: [HCEnvSection]
    private var dependencyEngine: DependencyEngine
    private var indexMap: [String: IndexPath]

    init() {
        let envSection = HCEnvBuilder.buildEnvSection()
        self.sections = [envSection]
        self.dependencyEngine = DependencyEngine(items: envSection.items)
        self.indexMap = [:]
        loadPersistedValues()
        rebuildIndexMap()
        rebuildDependencyEngine()
        refreshAllItems()
    }

    func item(at indexPath: IndexPath) -> HCCellItem {
        sections[indexPath.section].items[indexPath.row]
    }

    func updateItem(_ item: HCCellItem, value: Any?) -> [IndexPath] {
        item.value = value
        if item.type == .string || item.type == .stepper {
            item.detail = value.map { String(describing: $0) }
        }
        item.valueTransformer?(item)

        var changed = Set<String>()
        if !item.identifier.isEmpty {
            changed.insert(item.identifier)
            let propagated = dependencyEngine.propagateFromItemId(item.identifier)
            changed.formUnion(propagated)
        }

        persistIfNeeded(for: item)
        persistEnvConfig()

        var paths: [IndexPath] = []
        for itemId in changed {
            if let indexPath = indexMap[itemId] {
                paths.append(indexPath)
            }
        }
        return paths
    }

    func presentationForDisabledItem(_ item: HCCellItem) -> HCPresentationRequest {
        if let hint = item.disabledHint, !hint.isEmpty {
            return HCPresentationRequest.toast(message: hint)
        }
        return HCPresentationRequest.toast(message: "当前不可用")
    }

    private func loadPersistedValues() {
        for section in sections {
            for item in section.items {
                guard let storeKey = item.storeKey, !storeKey.isEmpty else {
                    continue
                }
                if let stored = UserDefaults.standard.object(forKey: storeKey) {
                    item.value = stored
                } else if let defaultValue = item.defaultValue {
                    item.value = defaultValue
                }
                if item.type == .string || item.type == .stepper {
                    item.detail = item.value.map { String(describing: $0) }
                }
            }
        }
    }

    private func rebuildDependencyEngine() {
        let items = sections.flatMap { $0.items }
        dependencyEngine = DependencyEngine(items: items)
    }

    private func refreshAllItems() {
        let itemsById = dependencyEngine.itemsById
        for section in sections {
            for item in section.items {
                item.recomputeBlock?(item, itemsById)
            }
        }
    }

    private func rebuildIndexMap() {
        var map: [String: IndexPath] = [:]
        for (sectionIndex, section) in sections.enumerated() {
            for (rowIndex, item) in section.items.enumerated() {
                if !item.identifier.isEmpty {
                    map[item.identifier] = IndexPath(row: rowIndex, section: sectionIndex)
                }
            }
        }
        indexMap = map
    }

    private func persistIfNeeded(for item: HCCellItem) {
        guard let storeKey = item.storeKey, !storeKey.isEmpty else {
            return
        }
        if let value = item.value {
            UserDefaults.standard.set(value, forKey: storeKey)
        } else {
            UserDefaults.standard.removeObject(forKey: storeKey)
        }
        UserDefaults.standard.synchronize()
    }

    private func persistEnvConfig() {
        let itemsById = dependencyEngine.itemsById
        guard let envItem = itemsById[HCEnvItemIdEnvType],
              let clusterItem = itemsById[HCEnvItemIdCluster],
              let isolationItem = itemsById[HCEnvItemIdIsolation],
              let versionItem = itemsById[HCEnvItemIdVersion] else {
            return
        }

        let config = HCEnvConfig()
        config.envType = HCEnvType(rawValue: intValue(envItem.value)) ?? .release
        config.clusterIndex = intValue(clusterItem.value)
        config.isolation = isolationItem.value as? String ?? ""
        config.version = versionItem.value as? String ?? "v1"
        HCEnvKit.saveConfig(config)
    }
}
